### ISSUE TITLE
Minibuffer completion width incorrect for some frameworks

### ISSUE THREAD

**Issue by jethrokuan** [org-roam/org-roam#2066]
Vertico's completion width is now halved:

I'm a little late, but this PR messes up Vertico's work, especially because Vertico always
runs in a minibuffer. This happens when the current window is halved and is smaller than
the `org-roam-node-display-template` requires. Perhaps bufferp current-buffer is not
working as intended.

Might be worth a try replace `(bufferp current-buffer)` with `(minibufferp)`?

```diff
-                         (1- (if (bufferp (current-buffer))
+                         (1- (if (minibufferp) 
```

My `org-roam-node-display-template`:

```elisp
(org-roam-node-display-template (concat "${title:*} " (propertize "${tags:100}" 'face 'org-tag)))
```

You can try variant paste this into `init.el`:

```diff
+  (defun my/org-roam-node-read--to-candidate (node template)
+    "Return a minibuffer completion candidate given NODE.
+  TEMPLATE is the processed template used to format the entry."
+    (let ((candidate-main (org-roam-node--format-entry
+                           template
+                           node
+                           (1- (if (minibufferp)
+                                   (window-width) (frame-width))))))
+      (cons (propertize candidate-main 'node node) node)))
+
+  (advice-add 'org-roam-node-read--to-candidate :override #'my/org-roam-node-read--to-candidate))
```

I would like this to be tried by users of different completion engines (Helm, vertico, whatever), because
this problem is repeated in a circle, every time :disappointed: .

_Originally posted by @PRESFIL in https://github.com/org-roam/org-roam/issues/2040#issuecomment-1023175509_

**Comment by PRESFIL** [comment]
Should I create an issue every time? Sorry if I just wanted to ask. :smile_cat:

**Comment by PRESFIL** [comment]
Since this is modification of #2040's PR, I would like @stuhlmueller to express his opinion.

**Comment by jethrokuan** [comment]
> Should I create an issue every time? Sorry if I just wanted to ask. 😸

@PRESFIL Yes, if it's not already in the list of issues. It's hard to track comments in PRs. Don't worry about creating too many issues, I can always close them. These issues are helpful for me to fix things later on too, when I find time.

**Comment by adql** [comment]
This happens to me on Ivy (Emacs Prelude) with the drop-in template
```emacs
(setq org-roam-node-display-template
      (concat "${title:*} "
              (propertize "${tags:10}" 'face 'org-tag)))
```
suggested in the [commit](https://github.com/org-roam/org-roam/commit/6f5d65abd9e707b3fdb10092a9fef3b739e143dd).

Strange that this bug hasn't gained more traction.

**Comment by PRESFIL** [comment]
I think the reason was this commit 001de3a (PR #2040) and not the one you link to.
Yes, it was older, but so often happens that those little bugs don't get noticed. That's
why I wanted @stuhlmueller to try what I suggested on my completeion-framework (helm).

This should determine the correct width for the minibuffer (frame-width) if it is a minibuffer,
or take the width of the current window if it is a window. I hope the minibuffer doesn't turn
vertical :smile: (haha, vertico is horizontal).

**Comment by PRESFIL** [comment]
Oh, I think now I understand. `org-roam-mode-read--completions` is called before the
`completion-read` is started, so the completion list created with the sizes of window
BEFORE the prompt is created, which can change sizes AFTER its launch.

It's incorrect to ask for the window size before running the prompt, but it's not
clear how to do it after the `completion-read` has started...

Checks inside `org-roam-node-read--to-candidate` It's not clear what they check ;)
The current window, which can be anything...

**Comment by hpfr** [comment]
Still relevant

**Comment by lnguyen4** [comment]
Can confirm that the fix (https://github.com/org-roam/org-roam/issues/2066#issue-1116478878) is still relevant on my emacs. This fix will make vertico's width of candidates display in the minibuffer extending to the whole frame width , when there are different windows on the frame.

**Comment by hpfr** [comment]
I think it’s impossible to fix this in the general case as long as we pre-format nodes based on user-controlled `org-roam-node-display-template` before handing them off to completing-read. The available width is calculated before `completing-read` is invoked, and we can’t predict the user’s completion interface.

As a quick example of current behavior, `org-roam-node-display-template` could be `"${title:*} ${tags:42}"`. The intent is that tags will start 42 characters from the right of the available width. The problem is that the template is processed for all nodes before completion begins, so the available width isn’t actually known. Currently, we effectively use `(window-width)` of the current window whenever `org-roam-node-find` is invoked (see https://github.com/org-roam/org-roam/pull/2040). But the current window width is often completely unrelated to the completion interface width.

The proposed fix is a lateral move and would break Helm. `(minibufferp)` is almost never true when `org-roam-node-find` is invoked (unless you invoke it when already in a minibuffer), and when `(minibufferp)` is true, `(window-width)` is the same as `(frame-width)`. The proposed fix effectively reverts to `(frame-width)`, which is not a true fix. You can also tell that the proposed fix is broken by activating `(vertico-buffer-mode)` with two side-by-side split windows. All the candidates will be too wide for the Vertico buffer, since `(frame-width)` is used.

We need to dig into `completing-read` and figure out if `org-roam--format-nodes-using-template` can be invoked “lazily” on each candidate, by passing it as a callback somehow. I think annotations serve this purpose. The problem is that the current behavior does not use annotations, so users probably expect completion to match against all the text in their `org-roam-node-display-template` by default (without having to rely on Orderless’s annotation matching, for example).

So maybe we need to parse the template twice? Once on all nodes, without a width, for the purposes of matching against user input, and again “lazily”, once we know the available width, for the purposes of display?

To my surprise, @minad actually worked on this code at one point (https://github.com/org-roam/org-roam/commit/3a54182bb1d637d32331c51eb2fa1a2f32dcb3c8) and is certainly an expert on idiomatic use of `completing-read`. I don’t know if you still use org-roam, but would you have any advice for us? Thank you!

The relevant code is ~150loc:
https://github.com/org-roam/org-roam/blob/903bd4ec56d29247990d005ed9052c201e18b812/org-roam-node.el#L525-L675

And the relevant custom variables:
https://github.com/org-roam/org-roam/blob/903bd4ec56d29247990d005ed9052c201e18b812/org-roam-node.el#L35-L128

**Comment by minad** [comment]
@hpfr I don't use org-roam anymore - it has been a while. Generally reformatting candidates dynamically/lazily is not well supported by completing-read. Two things you could try:

1. Use dynamically computed widths in display spaces. For example `consult--annotate-align` uses this technique to recompute alignment from the variable `consult--annotate-align-width`.

2. Reformat candidates via the `:affixation-function`. I use this in `consult-line` to lazily highlight lines, see `consult--line-fontify`. There the candidate string itself is not changed, only face properties are added. So if you really rewrite the candidate strings, it may create problems. Also make sure that you don't lose the completion style highlighting if you rewrite the candidate.

In the end the completion system you use matters, and you should test that it works correctly. For my projects I decided to rely only on systems which are fully compliant with the default UI and dynamic completion tables (Vertico, Icomplete, Mct).

**Comment by hpfr** [comment]
Thank you! Those look like useful examples. Display properties seem much better suited to alignment than formatting literal spaces into the candidates.
